compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs

```diff
@@ -187,55 +187,62 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 err.span_label(self.tcx.def_span(def_id), &format!("{} defined here", found));
             }
         } else if !self.check_for_cast(err, expr, found, expected, expected_ty_expr) {
-            let is_struct_pat_shorthand_field =
-                self.maybe_get_struct_pattern_shorthand_field(expr).is_some();
+            let struct_pat_shorthand_field = self.maybe_get_struct_pattern_shorthand_field(expr);
             let methods = self.get_conversion_methods(expr.span, expected, found, expr.hir_id);
             if !methods.is_empty() {
-                if let Ok(expr_text) = self.sess().source_map().span_to_snippet(expr.span) {
-                    let mut suggestions = iter::zip(iter::repeat(&expr_text), &methods)
-                        .filter_map(|(receiver, method)| {
-                            let method_call = format!(".{}()", method.name);
-                            if receiver.ends_with(&method_call) {
-                                None // do not suggest code that is already there (#53348)
+                let mut suggestions = iter::zip(iter::repeat(&expr), &methods)
+                    .filter_map(|(receiver_expr, method)| {
+                        let method_call = format!(".{}()", method.name);
+                        fn method_ident(expr: &hir::Expr<'_>) -> Option<Ident> {
+                            match expr.kind {
+                                ExprKind::MethodCall(receiver_method, ..) => Some(receiver_method.ident),
+                                ExprKind::Unary(_, expr) | ExprKind::AddrOf(.., expr) => method_ident(expr),
+                                _ => None
+                            }
+                        }
+                        let method_ident = method_ident(&receiver_expr);
+                        if let Some(method_ident) = method_ident
+                            && method_ident.name == method.name
+                        {
+                            None // do not suggest code that is already there (#53348)
```

Comment:
This is an early exit from the `filter_map`, isn't it?  Could we just `return None` and avoid the rightward drift below?